### ISSUE TITLE
Enable description of client/subclient hierarchy

### ISSUE THREAD

**Issue by annelo-msft** [microsoft/cadl#656]
We will need a way to generate clients and subclients from Cadl as described in the guidelines: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients.

**Comment by daviwil** [comment]
Here's a writeup of some design thoughts while looking at a few of the issues @annelo-msft filed:


# Table of Contents

1.  [Requirements](#org98e94f6)
    1.  [Resource Clients](#orge30ae05)
    2.  [Operation Group Clients](#org19f77d9)
2.  [Multiple services per Cadl "document"](#org89bf0f3)
    1.  [Service client parameters](#org145d634)



<a id="org98e94f6"></a>

# Requirements

The request for subclient support is here: <https://github.com/microsoft/cadl/issues/656>

The desired design for clients and subclients in C# is described here: <https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-subclients>

There are a few different kinds of subclients but only two sound relevant to Cadl itself:

-   **Resource clients** for methods related to a particular resource containing some information about the resource
-   **Operation group clients** which are grouped by a particular name

Whatever we do to facilitate the generation of these clients should be minimal and not tightly focused on the needs of a specific language.

There are also **Operation** and **Pageble** clients, but those are just classes returned from long-running or pageable collection operations which follow standard patterns.

We also need to make it possible for multiple top-level service clients to be defined


<a id="orge30ae05"></a>

## Resource Clients

For resource clients, there needs to be a way to initialize a client based on a particular resource type and identifier.

In Cadl, we have the ability to define specific resource lifecycle operations for a given resource type either by reusing standard interfaces or operation signatures or by using the \`@readsResource\`, \`@createsResource\` etc decorators directly on any operation.

I don't believe we need to provide any additional functionality in Cadl's resource modelling to enable the generation of resource clients.  What we **can** do, however, is provide a unified view of all operations associated with a given resource type and a list of all resource types known to have operations.

Prior art for this can be found in the `cadl-azure-resource-manager` library's `getArmResources` function which returns an object with an `operations` property providing the following information:

    
    export interface ArmResourceOperations {
      lifecycle: ArmLifecycleOperations;
      lists: { [key: string]: ArmResourceOperation };
      actions: { [key: string]: ArmResourceOperation };
    }
    
    export interface ArmLifecycleOperations {
      read?: ArmResourceOperation;
      createOrUpdate?: ArmResourceOperation;
      update?: ArmResourceOperation;
      delete?: ArmResourceOperation;
    }

If `cadl-azure-core` provides a similar API then DPG client generators will be able to use it to generate resource clients with whatever conventions required for a particular language.


<a id="org19f77d9"></a>

## Operation Group Clients

For operation group clients, we already have the ability to mark interfaces and namespaces with a name which can be used in Swagger output as operation groups: the `@tag` decorator.

However, `@tag` isn't really appropriate for this purpose, so we would need one of the following:

-   A more appropriately-named decorator for adding operations and containers of operations to named operation groups (which doesn't really feel right)
-   A decorator to mark a namespace or interface as containing a set of related operations
-   A convention for emitters to use to automatically recognize a namespace or interface as an inherent operation group

The third option would be the best, but how do we do it?


<a id="org89bf0f3"></a>

# Multiple services per Cadl "document"

We've also got a request for defining multiple services in one larger document:

<https://github.com/microsoft/cadl/issues/678>

To enable this functionality, we would need the following (and possibly more):

-   The `server/service` decorators would need to allow more than one service to be defined in a total document
-   Route-scanning logic would need to be service-aware and associate routes with the correct service
-   Resource operation logic would also need to be service-aware so that only resources and related operations defined in a specific service are returned

Basically, any decorator or code that is storing information about a service **must** resolve the service namespace and use it as a "parent key" for storing its information.

We could provide some helper logic to automate this process for service-oriented decorators so that the decorator author doesn't have to think about it.


<a id="org145d634"></a>

## Service client parameters

This is a necessary requirement for generating good client SDKs: <https://github.com/microsoft/cadl/issues/664>

Since we already have the `@server` decorator, this would be the ideal approach for identifying client-level parameters since they have an impact on the URI parameters used when invoking operations.

We just need to ensure that this approach will cover all known scenarios for data plane services.

**Comment by markcowl** [comment]
Next steps from design meeting:
- [ ] Create a decorator for client grouping that can be applied on the namespace, interface, and operation level.  
  - Use an enumeration over the names of defined clients
  - The decorator references a client enum value to associate the operations in the decorated item with a particular client
- [ ] Create a decorator that can be applied to a root namespace that specifies a client
  - Decorator defines construction parameters
  - Need to determine the interaction with `@server` decorators
    - Perhaps reuse server parameter models
    - Perhaps make Client a model template that can be decorated with @server parameters
- [ ] Support using a sidecar file for more complex client definitions
  - See: https://teams.microsoft.com/l/message/19:16f8f1ed70d044139be153707bdaee0e@thread.skype/1658416290857?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1658416290857&teamName=Azure%20SDK&channelName=Cadl%20Engineering&createdTime=1658416290857  for details/example
  - This would allow more complex modeling of operations referenced from the source doc
  - Should eventually have linting rules for things like unreferenced operations to help with maintenance
- Decorators are used when possible for simple grouping of operations, sidecar file is used when finer grained control or more complex mapping are needed
- [ ] Model a few data plane services and revisit

**Comment by daviwil** [comment]
Implementation issue: https://github.com/microsoft/cadl/issues/778

**Comment by markcowl** [comment]
reopening for revisit

**Comment by daviwil** [comment]
@markcowl Since we have #1786 now, do we need this issue any longer?

**Comment by markcowl** [comment]
Closing in favor of #799  as the rest is implemented
